Auth basique

onora-hbl opened the pull request:
La base du système d'auth est fonctionnelle.

Il manque:
- La vérification des retour d'erreur
- Le renew des refresh_token a chaque utilisation


Exemple d'utilisation de l'auth dans le front:
```ts
function MyComponent() {
    const userContext = React.useContext(UserContext);

    useEffect(() => {
        const token = userContext.getAccessToken();
        fetch(BACK_URL + "/api/endpoint", {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
                Authorization: "Bearer " + token,
            }});
    });
}
```

Exemple d'utilisation de l'auth dans le back:
```ts
@Get('test')
@UseGuards(AuthGuard)
test(@CurrentUser() user: User) {
    return 'Hey ' + user.login;
}
```

onora-hbl committed:
feat(AUTH): Add basic auth redirections to 42 and callback

Simple redirect to 42 when /auth is getted

Get the user on 42 callback
and send it to front

onora-hbl committed:
fix(MODULES): Make AppModule global

As a global module, the AppModule can provide services to all submodules

onora-hbl committed:
feat(DB): Add Prisma support

Setup a default prisma connection and schema

onora-hbl committed:
feat(AUTH): Add basic functions to get users

Basic example to test if the user already exists on log

onora-hbl committed:
feat(USERS): Add basic registration process

Save in database the user using the LoggedUser

onora-hbl committed:
refactor(AUTH): Make all endpoints as Rest

Remove redirection of users from the api to the front

Replace it with
Rest API endpoints to make the token exchange

onora-hbl committed:
fix(AUTH): Change error message

Change to a more accurate error message

onora-hbl committed:
Merge branch 'main' into auth

onora-hbl committed:
feat(ROUTER): Add basic router logic

onora-hbl committed:
fix(BACK): Enable CORS

Enable cors to test

onora-hbl committed:
feat(AUTH): Add basic token generation

Add basic functions to generate user tokens

onora-hbl committed:
feat(LOG): Implement simple context to manage login

Create a simple context to store all useful informations of the login
logic

onora-hbl committed:
refactor(UserContext): Move UserContext to specific file

onora-hbl committed:
feat(AUTH): Implement basic token refreshing

Simple implementation for token refreshing, need to add the modification
of pesonal value

onora-hbl committed:
feat(REFRESH): Refreshing of token

Refresh the access token when access token expire

onora-hbl committed:
feat(AUTH): Add AuthGuard

Add the guard used to verifiy if the use is logged

onora-hbl committed:
feat(AUTH): Add CurrentUser decorator

Add a decorator to get the user in an endpoint protected by the
AuthGuard

jerem-ma reviewed (changes requested):
Super code. Complexe à comprendre à comprendre pour un néophyte comme moi, mais ça a l'air magique et de fonctionner à merveille. Pense bien à remplacer ta fonction magique de 42.js de `processes` à `process` parce que bahh, t'as qu'un seul process quoi, c'est un peu con.
Bisoux.

jerem-ma reviewed (commented):
remove moi plouvel T.T

jerem-ma commented on back/.env.sample:
```diff
@@ -1,4 +1,14 @@
+DATABASE_PASSWORD=
+DATABASE_NAME=
+
+DATABASE_URL=postgresql://${DATABASE_USER}:${DATABASE_PASSWORD}@${DATABASE_HOST}:${DATABASE_PORT}/${DATABASE_NAME}
```
pouet

jerem-ma commented on back/package.json:
```diff
@@ -23,11 +23,18 @@
+        "@types/passport-jwt": "^3.0.8",
+        "42.js": "^1.2.4",
+        "class-transformer": "^0.5.1",
+        "class-validator": "^0.14.0",
```
pouet n o 2

jerem-ma reviewed (dismissed):
remove moi plouvel T.T

jerem-ma commented:
Désolé plouvel, je pensais qu'on pouvait delete :'(

onora-hbl committed:
style(AUTH): Change processes to process

onora-hbl committed:
refactor(UserContext): Remove not necessary code

Remove default value of UserContext because it is never used without a
provider

onora-hbl commented:
(Jerem m a autorisé a merge ma PR)
